### PR TITLE
Ogury - 1.3.5.0

### PR DESCRIPTION
**Pull request by DamienRivetOgury**
# Summary

Add support for **Interstitial**, **Rewarded Video** and **Banner** from **Ogury**.

### CODE REVIEW

File: Ogury/OguryRewardedVideoCustomEvent.m [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    if (item.rewardName != nil && ![item.rewardName isEqualToString:@""]) {
+        currencyType = item.rewardName;
+    }
+    
+    if (item.rewardValue != nil && ![item.rewardValue isEqualToString:@""]) {
+        amount = item.rewardValue.integerValue;
+    }
+
+    MPRewardedVideoReward *reward = [[MPRewardedVideoReward alloc] initWithCurrencyType:currencyType amount:@(amount)];
```
Comment (chauduyphanvu) [inline Ogury/OguryRewardedVideoCustomEvent.m:109]: `MPReward`

File: Ogury/OguryRewardedVideoCustomEvent.m [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    [self.delegate fullscreenAdAdapter:self didFailToLoadAdWithError:error];
+}
+
+- (void)oguryAdsOptinVideoAdRewarded:(OGARewardItem *)item {
+    NSString *currencyType = kMPRewardCurrencyTypeUnspecified;
+    NSInteger amount = kMPRewardCurrencyAmountUnspecified;
+
+    if (item.rewardName != nil && ![item.rewardName isEqualToString:@""]) {
+        currencyType = item.rewardName;
+    }
```
Comment (chauduyphanvu) [inline Ogury/OguryRewardedVideoCustomEvent.m:103]: Nil check for `item` first?

File: Ogury/OguryRewardedVideoCustomEvent.m [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+- (void)oguryAdsOptinVideoAdError:(OguryAdsErrorType)errorType {
+    NSError *error = [NSError ogy_MoPubErrorFromOguryError:errorType];
+
+    MPLogAdEvent([MPLogEvent adShowFailedForAdapter:NSStringFromClass([self class]) error:error], self.adUnitId);
+
+    [self.delegate fullscreenAdAdapter:self didFailToLoadAdWithError:error];
+}
+
+- (void)oguryAdsOptinVideoAdLoaded {
+    [self.delegate fullscreenAdAdapterDidLoadAd:self];
```
Comment (chauduyphanvu) [inline Ogury/OguryRewardedVideoCustomEvent.m:84]: Add log message.

File: Ogury/OguryBannerCustomEvent.m [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    OguryAdsBannerSize *sizeOguryBanner = [OguryBannerCustomEvent getOgurySize:size];
+
+    if (!sizeOguryBanner) {
+        [self.delegate inlineAdAdapter:self didFailToLoadAdWithError:[NSError errorWithCode:MOPUBErrorNoInventory]];
+        return;
+    }
+
+    [OguryAdapterConfiguration applyTransparencyAndConsentStatusWithParameters:info];
+
+    self.adUnitId = info[kOguryConfigurationAdUnitId];
```
Comment (chauduyphanvu) [inline Ogury/OguryBannerCustomEvent.m:40]: Validate ad unit ID and fail if applicable.

File: Ogury/OguryAdapterConfiguration.m [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+#pragma mark - Methods
+
++ (void)updateInitializationParameters:(NSDictionary *)parameters {
+    // Not implemented
+}
+
+- (void)initializeNetworkWithConfiguration:(NSDictionary<NSString *, id> * _Nullable)configuration complete:(void(^ _Nullable)(NSError * _Nullable))complete {
+    [[OguryAds shared] defineMediationName:OguryConfigurationMediationName];
+    
+    NSString *assetKey = configuration[OguryConfigurationKeyAssetKey];
```
Comment (chauduyphanvu) [inline Ogury/OguryAdapterConfiguration.m:52]: Nil check on `configuration`.

File: Ogury/NSError+Ogury.m [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+#import "NSError+Ogury.h"
+#import "OguryAdapterConfiguration.h"
+
+@implementation NSError (Ogury)
+
++ (NSError *)ogy_MoPubErrorFromOguryError:(OguryAdsErrorType)oguryError {
+    NSNumber *mopubErrorCode;
+    NSString *localizedDescription;
+
+    switch (oguryError) {
```
Comment (chauduyphanvu) [inline Ogury/NSError+Ogury.m:14]: Please move this to the `AdapterConfiguration`. No need for a separate class.

**Review by chauduyphanvu — CHANGES_REQUESTED** [review]
need additional nil checks, error handlings, and log messages/event

File: Ogury/NSError+Ogury.m (hunk 6 above)
Comment (DamienRivetOgury) [reply Ogury/NSError+Ogury.m]: Moved in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryAdapterConfiguration.m (hunk 5 above)
Comment (DamienRivetOgury) [reply Ogury/OguryAdapterConfiguration.m:64]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryRewardedVideoCustomEvent.m (hunk 3 above)
Comment (DamienRivetOgury) [reply Ogury/OguryRewardedVideoCustomEvent.m:110]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryRewardedVideoCustomEvent.m (hunk 2 above)
Comment (DamienRivetOgury) [reply Ogury/OguryRewardedVideoCustomEvent.m]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryInterstitialCustomEvent.m [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+- (void)requestAdWithAdapterInfo:(NSDictionary *)info adMarkup:(NSString *)adMarkup {
+    [OguryAdapterConfiguration applyTransparencyAndConsentStatusWithParameters:info];
+
+    self.adUnitId = info[kOguryConfigurationAdUnitId];
+
+    self.interstitial = [[OguryAdsInterstitial alloc] initWithAdUnitID:self.adUnitId];
+    self.interstitial.interstitialDelegate = self;
+
+    [self.interstitial load];
```
[the remark this answers is not in the document]
Comment (DamienRivetOgury) [reply Ogury/OguryInterstitialCustomEvent.m:46]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryInterstitialCustomEvent.m [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+- (BOOL)hasAdAvailable {
+    return self.interstitial.isLoaded;
+}
+
+- (void)presentAdFromViewController:(UIViewController *)viewController {
+    if (self.interstitial.isLoaded) {
+        [self.delegate fullscreenAdAdapterAdWillAppear:self];
+        [self.interstitial showAdInViewController:viewController];
+    }
```
[the remark this answers is not in the document]
Comment (DamienRivetOgury) [reply Ogury/OguryInterstitialCustomEvent.m]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryRewardedVideoCustomEvent.m [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+#import <OguryAds/OguryAds.h>
+#import "OguryAdapterConfiguration.h"
+#import "NSError+Ogury.h"
+
+@interface OguryRewardedVideoCustomEvent () <OguryAdsOptinVideoDelegate>
+
+#pragma mark - Properties
+
+@property (nonatomic, copy) NSString *adUnitId;
+@property (nonatomic, strong) OguryAdsOptinVideo *optinVideo;
```
[the remark this answers is not in the document]
Comment (DamienRivetOgury) [reply Ogury/OguryRewardedVideoCustomEvent.m]: Changed in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryRewardedVideoCustomEvent.m [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+- (void)requestAdWithAdapterInfo:(NSDictionary *)info adMarkup:(NSString *)adMarkup {
+    [OguryAdapterConfiguration applyTransparencyAndConsentStatusWithParameters:info];
+
+    self.adUnitId = info[kOguryConfigurationAdUnitId];
+    self.optinVideo = [[OguryAdsOptinVideo alloc] initWithAdUnitID:self.adUnitId];
+    self.optinVideo.optInVideoDelegate = self;
+
+    [self.optinVideo load];
```
[the remark this answers is not in the document]
Comment (DamienRivetOgury) [reply Ogury/OguryRewardedVideoCustomEvent.m]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryRewardedVideoCustomEvent.m [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+- (BOOL)isRewardExpected {
+    return YES;
+}
+
+- (BOOL)hasAdAvailable {
+    return self.optinVideo.isLoaded;
```
[the remark this answers is not in the document]
Comment (DamienRivetOgury) [reply Ogury/OguryRewardedVideoCustomEvent.m]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryRewardedVideoCustomEvent.m [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+- (void)presentAdFromViewController:(UIViewController *)viewController {
+    if (self.optinVideo.isLoaded) {
+        [self.optinVideo showAdInViewController:viewController];
+    }
```
[the remark this answers is not in the document]
Comment (DamienRivetOgury) [reply Ogury/OguryRewardedVideoCustomEvent.m]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryRewardedVideoCustomEvent.m (hunk 1 above)
Comment (DamienRivetOgury) [reply Ogury/OguryRewardedVideoCustomEvent.m]: Changed in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryAdapterConfiguration.m [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+    if (!assetKey || [assetKey isEqualToString:@""]) {
+        NSError *error = [NSError errorWithDomain:kOguryErrorDomain
+                                             code:MOPUBErrorAdapterInvalid
+                                         userInfo:@{NSLocalizedDescriptionKey:@"OguryAdsAssetKeyNotValidError. An error occurred during the initialization of the SDK."}];
```
[the remark this answers is not in the document]
Comment (DamienRivetOgury) [reply Ogury/OguryAdapterConfiguration.m]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryBannerCustomEvent.m [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+- (void)dealloc {
+    self.banner.bannerDelegate = nil;
+}
+
+- (void)requestAdWithSize:(CGSize)size adapterInfo:(NSDictionary *)info adMarkup:(NSString *)adMarkup {
+    OguryAdsBannerSize *sizeOguryBanner = [OguryBannerCustomEvent getOgurySize:size];
+
+    if (!sizeOguryBanner) {
+        [self.delegate inlineAdAdapter:self didFailToLoadAdWithError:[NSError errorWithCode:MOPUBErrorNoInventory]];
```
[the remark this answers is not in the document]
Comment (DamienRivetOgury) [reply Ogury/OguryBannerCustomEvent.m]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryBannerCustomEvent.m [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+    [self.delegate inlineAdAdapterDidTrackClick:self];
+}
+
+- (void)oguryAdsBannerAdClosed:(OguryAdsBanner *)bannerAds {
+     [self.delegate inlineAdAdapterDidEndUserAction:self];
+}
+
+- (void)oguryAdsBannerAdDisplayed:(OguryAdsBanner *)bannerAds {
+     [self.delegate inlineAdAdapterDidTrackImpression:self];
+}
```
[the remark this answers is not in the document]
Comment (DamienRivetOgury) [reply Ogury/OguryBannerCustomEvent.m]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryInterstitialCustomEvent.m [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+
+    [self.interstitial load];
+}
+
+- (BOOL)isRewardExpected {
+    return NO;
+}
+
+- (BOOL)hasAdAvailable {
+    return self.interstitial.isLoaded;
```
[the remark this answers is not in the document]
Comment (DamienRivetOgury) [reply Ogury/OguryInterstitialCustomEvent.m]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

**Comment by DamienRivetOgury** [comment]
Hi @chauduyphanvu,

Regarding AdUnitId validation interstitial ads, for some very specific cases we do accept an empty value.

File: Ogury/OguryBannerCustomEvent.m (hunk 4 above)
Comment (DamienRivetOgury) [reply Ogury/OguryBannerCustomEvent.m]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryRewardedVideoCustomEvent.m [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+#pragma mark - Methods
+
+- (void)dealloc {
+    self.optinVideo.optInVideoDelegate = nil;
+}
+
+- (void)requestAdWithAdapterInfo:(NSDictionary *)info adMarkup:(NSString *)adMarkup {
+    [OguryAdapterConfiguration applyTransparencyAndConsentStatusWithParameters:info];
+
+    self.adUnitId = info[kOguryConfigurationAdUnitId];
```
[the remark this answers is not in the document]
Comment (DamienRivetOgury) [reply Ogury/OguryRewardedVideoCustomEvent.m]: Added in [`e149bb4`](https://github.com/mopub/mopub-ios-mediation/pull/388/commits/e149bb4972c186365974ed36add750fdcca16daa).

File: Ogury/OguryInterstitialCustomEvent.m [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+#pragma mark - Methods
+
+- (void)dealloc {
+    self.interstitial.interstitialDelegate = nil;
+}
+
+- (void)requestAdWithAdapterInfo:(NSDictionary *)info adMarkup:(NSString *)adMarkup {
+    [OguryAdapterConfiguration applyTransparencyAndConsentStatusWithParameters:info];
+
+    self.adUnitId = info[kOguryConfigurationAdUnitId];
```
[the remark this answers is not in the document]
Comment (DamienRivetOgury) [reply Ogury/OguryInterstitialCustomEvent.m:29]: Hi @chauduyphanvu,

Regarding ad unit ID validation interstitial ads (and only for this format), for some very specific cases we do accept an empty value.

**Comment by DamienRivetOgury** [comment]
Hi @chauduyphanvu, you should be able to clone the fork from our organization. If it's not the case, I'll have a look.

**Comment by chauduyphanvu** [comment]
Done in https://github.com/mopub/mopub-ios-mediation/commit/da057748d15134d96c99f49e454883ad901dd779. Thanks!

**Pull request closed without merging**